### ISSUE TITLE
False positive "keyboard-interactive" / "Further authentication required" from Multi-Factor = no connection

### ISSUE THREAD

**Issue by AlbertMN** [phpseclib/phpseclib#1795]
Hey there. So, the trail of this issue is pretty long, but after hours of debugging, I've found that this library, in combination with the Multi-Factor setup on my server, makes me unable to connect to SSH via. PHP, using [Flysystem](https://github.com/thephpleague/flysystem) which uses your library to make SSH connections _(or at least the authentication part?)_

**Issue;**
My understanding is, that my Ubuntu server gives a "false positive" to your library about needing multi-factor authentication...
So, as I've set up multi-factor for some users on the server, it shows a brief `Further authentication required` when SSH connecting _(also via [PuTTY](https://putty.org/))_, but it quickly allows the user to pass anyway.

I'm using the [Google Authenticator PAM module](https://github.com/google/google-authenticator-libpam) library, and also made an issue about the false-positive message here; https://github.com/google/google-authenticator-libpam/issues/214

So, in the end it seems to be a mix of the fact that the message, which probably _(hopefully)_ shouldn't be there, claiming authentication is required, and the fact that this library is a bit strict _(probably rightfully so)_ about this, and then provides no way to establish the connection.

**What I have tried;**
I have confirmed that it all works on a server that doesn't have multi-factor on for any user, and thereby doesn't show the message. Furthermore, I have tried to just pass a fake multi-factor code to your library to see if that helps, but with no luck _(my guess is, that after the code is given to the server, it expects the server to tell if it was successful or failed, but as the server isn't really expecting a code, it gets no known response?)_.

Let me know if I've gotten anything wrong - I know it's a bit of a stretch to make this an issue for this library, but my thought was, that if SSH terminals like PuTTY can handle it, I was hoping there's a way.

Thanks, and thanks for all your hard work!

**Comment by terrafrost** [comment]
Also, can you post the code you're using? Replace any sensitive info (keys or passwords) with dummy sensitive info, obviously, but knowing how you're calling the login method would help.

**Comment by AlbertMN** [comment]
@terrafrost thanks for the swift reply! And sorry for the delay on my end.

I see I said "SSH connection" earlier, but it's actually SFTP*

So, this is where it gets a bit tricky... I'm using the Flysystem PHP library to connect, so I don't directly do any phpseclib code.

But, I went down the rabbit hole and at least got the log output after connection fails; https://pastebin.mozilla.org/MbL7788g

The code from the `SftpConnectionProvider.php` file in the Flysystem SFTP extension (https://github.com/thephpleague/flysystem-sftp), which handles the connection, is as follows; (https://github.com/thephpleague/flysystem-sftp/blob/2.x/SftpConnectionProvider.php#L124)
```
$connection = new SFTP($this->host, $this->port, $this->timeout);
$connection->disableStatCache();

try {
    $this->checkFingerprint($connection);
    $this->authenticate($connection);
} catch (Throwable $exception) {
    $connection->disconnect();
    throw $exception;
}
```
And it fails on `throw $exception`. 

I have just tested, and re-confirmed, that everything works, if I just disable multi-factor authentication on my server completely, but the fact is that the user I'm attempting to connect to, never _actually_ requires multi-factor, but it still, as seen in the log, gets `keyboard-interactive` from the server.

**Comment by terrafrost** [comment]
> I see I said "SSH connection" earlier, but it's actually SFTP*
>
> So, this is where it gets a bit tricky... I'm using the Flysystem PHP library to connect, so I don't directly do any phpseclib code.
>
> But, I went down the rabbit hole and at least got the log output after connection fails; https://pastebin.mozilla.org/MbL7788g

The pastebin link returns an error 404 for me. In lieu of being able to see what you posted there...

Authentication takes place at the SSH layer so I'd still like to see the SSH logs - not the SFTP ones.

If you can't use `echo $ssh->getLog()` then you can use `define('NET_SSH2_LOGGING', 3);`. That'll enable real time logging.

> The code from the SftpConnectionProvider.php file in the Flysystem SFTP extension (https://github.com/thephpleague/flysystem-sftp), which handles the connection, is as follows; (https://github.com/thephpleague/flysystem-sftp/blob/2.x/SftpConnectionProvider.php#L124)

Can you post the code _you_ are using to login? I understand that that code isn't going to be pure-phpseclib and that's fine - I just want to see how you're trying to call it atm. Once I see it we can proceed from there.

Thanks!

**Comment by AlbertMN** [comment]
@terrafrost of course :)

```
$adapter = new SftpAdapter(new SftpConnectionProvider("XX.XX.XX.XX", // host (required)
  "localuser", //username (required)
  null, //password (optional, default: null) set to null if privateKey is used
  "/var/keys/priv-key", // private key 
), "/root/path"); // root path (required)

$filesystem = new \League\Flysystem\Filesystem($adapter)
```

Taken pretty straight out of the example from the Flysystem docs.

**Comment by terrafrost** [comment]
So `setupConnection` calls `$this->authenticate($connection)`, which does this:

```php
        if ($this->privateKey !== null) {
            $this->authenticateWithPrivateKey($connection);
        } elseif ($this->useAgent) {
            $this->authenticateWithAgent($connection);
        } elseif ( ! $connection->login($this->username, $this->password)) {
            throw UnableToAuthenticate::withPassword();
        }
```
`authenticateWithPrivateKey` does this:

```php
        $privateKey = $this->loadPrivateKey();

        if ($connection->login($this->username, $privateKey)) {
            return;
        }

        if ($this->password !== null && $connection->login($this->username, $this->password)) {
            return;
        }

        throw UnableToAuthenticate::withPrivateKey();
```
So it looks like Flysystem supports multi-factor authentication.

Could be that your key isn't in a supported format. Could be that the server is expecting the password first instead of the private key first. I'd still be interested in seeing the SSH logs. Try `define('NET_SSH2_LOGGING', 3);` if `echo $ssh->getLog()` didn't work.

**Comment by AlbertMN** [comment]
@terrafrost correct, up until the point where I don't know if we're 100% on the same page; multi factor _does_ work, and not using multi-factor _(just password)_ also works, and yeah, Flysystem supports both. But, the issue here is, that as soon as I enable multi-factor for just one user on my server, there's somewhat an "artifact" of multi-factor on all other users too; so, here I'm trying to log in with a user that doesn't actually have multi factor, but as seen in the _(now disappeared)_ log; when connecting, it says _"keyboard-interactive"_, which prompts phpseclib to _think_ it needs multi-factor; however, it actually logs in right after, but phpseclib terminates the connection, thinking a code is required when it has none. _(also trying with a random code doesn't work either)_.

Apologies if you understood this already.

I didn't try getting SSH logging, as the "SFTP" class was used - do you still wish the SSH log, and not just SFTP log?

**Comment by terrafrost** [comment]
Ah - ok - I guess I did misunderstand.

Anyway, SSH logs would still help. And yes - I want SSH vs SFTP. Authentication doesn't take place at the SFTP layer - it takes place at the SSH layer. And because you're not being authenticated the SFTP logs should be empty anyway. Further, if I really wanted, I could extract the SFTP logs from the SSH2 logs - I can't go the other direction from the SFTP logs.
